external/storm-kafka-new-consumer-api/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,457 @@
+            }
+            acked.get(tp).add(msgId);
+        } finally {
+            ackCommitLock.unlock();
+        }
+    }
+
+    // ======== Fail =======
+
+    @Override
+    public void fail(Object messageId) {
+        final org.apache.storm.kafka.spout.MessageId msgId = (org.apache.storm.kafka.spout.MessageId) messageId;
+
+        if (isInBlackList(msgId)) {
+            removeFromBlacklist(msgId);
+        } else {
+            final TopicPartition tp = msgId.getTopicPartition();
+            // limit to max number of retries
+            if (msgId.numFails() >= maxRetries) {
+                LOG.debug("Reached the maximum number of retries. Adding [{]} to list of messages to be committed to kafka", msgId);
+                ack(msgId);
+                removeFromFailed(tp, msgId);
+            } else {
+                addToFailed(tp, msgId);
+            }
+        }
+    }
+
+    private void addToFailed(TopicPartition tp, org.apache.storm.kafka.spout.MessageId msgId) {
+        if (!failed.containsKey(tp)) {
+            failed.put(tp, new HashSet<org.apache.storm.kafka.spout.MessageId>());
+        }
+        final Set<org.apache.storm.kafka.spout.MessageId> msgIds = failed.get(tp);
+        if (msgIds.contains(msgId)) {      // do this to update the counter of the message
+            msgIds.remove(msgId);
+        }
+        msgId.incrementNumFails();        // increment number of failures counter
+        msgIds.add(msgId);
+    }
+
+    private void removeFromFailed(TopicPartition tp, org.apache.storm.kafka.spout.MessageId msgId) {
+        if (failed.containsKey(tp)) {
+            final Set<org.apache.storm.kafka.spout.MessageId> msgIds = failed.get(tp);
+            msgIds.remove(msgId);
+            if (msgIds.isEmpty()) {
+                failed.remove(tp);
+            }
+            LOG.debug("Removing from failed list [{}, {}]", tp, msgId);
+        }
+    }
+
+    // ======== Activate / Deactivate =======
+
+    @Override
+    public void activate() {
+        // Shouldn't have to do anything for now. If specific cases need to be handled logic will go here
+    }
+
+    @Override
+    public void deactivate() {
+        commitAckedTuples();
+    }
+
+    @Override
+    public void close() {
+        try {
+            kafkaConsumer.wakeup();
+            commitAckedTuples();
+        } finally {
+            //remove resources
+            kafkaConsumer.close();
+        }
+    }
+
+    @Override
+    public void declareOutputFields(OutputFieldsDeclarer declarer) {
+        declarer.declareStream(kafkaSpoutStream.getStreamId(), kafkaSpoutStream.getOutputFields());
+    }
+
+    // ====== Private helper methods ======
+
+    private void commitAckedTuples() {
+        final Map<TopicPartition, OffsetAndMetadata> toCommitOffsets = new HashMap<>();
+
+        LOG.debug("Committing acked offsets to Kafka");
+        // lock because ack and commit happen in different threads
```

Comment:
Only when the rebalance happens.  And even then we don't really need a separate lock object, we could just make this method synchronized.